`TestFetch` from `pkg/git` modify global git config

# Expected Behavior

Running `go test ./...` doesn't modify my personal / global git configuration.

# Actual Behavior

Running the test on a system where the git config is read-only (for example NixOS), the test fails.

```
--- FAIL: TestFetch (0.20s)
    --- FAIL: TestFetch/test-good (0.05s)
        git_test.go:342: Fetch() error = exit status 255, wantErr false
        git_test.go:423: log message: 'Successfully cloned'
             should contain: 'Error running git [config --add --global safe.directory /tmp/TestFetchtest-good4173527465/002]: exit status 255
            error: could not lock config file /home/vincent/.config/git/config: Read-only file system
            '
        git_test.go:323: ["error" git]: Error running git [config --add --global safe.directory /tmp/TestFetchtest-good4173527465/002]: exit status 255
            error: could not lock config file /home/vincent/.config/git/config: Read-only file system
    --- FAIL: TestFetch/test-clone-with-sparse-checkout (0.04s)
        git_test.go:342: Fetch() error = exit status 255, wantErr false
        git_test.go:350: Unable to read sparse-checkout file
        git_test.go:323: ["error" git]: Error running git [config --add --global safe.directory /tmp/TestFetchtest-clone-with-sparse-checkout2075723561/002]: exit status 255
            error: could not lock config file /home/vincent/.config/git/config: Read-only file system
    --- FAIL: TestFetch/test-clone-with-submodules (0.11s)
        git_test.go:342: Fetch() error = exit status 255, wantErr false
        git_test.go:323: ["error" git]: Error running git [config --add --global safe.directory /tmp/TestFetchtest-clone-with-submodules367881319/003]: exit status 255
            error: could not lock config file /home/vincent/.config/git/config: Read-only file system
panic: runtime error: index out of range [1] with length 1 [recovered]
	panic: runtime error: index out of range [1] with length 1

goroutine 32 [running]:
testing.tRunner.func1.2({0x67da00, 0xc000027968})
	/nix/store/njs424yyxxjhxibkpx5m36flr26nzzn6-go-1.17.8/share/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/nix/store/njs424yyxxjhxibkpx5m36flr26nzzn6-go-1.17.8/share/go/src/testing/testing.go:1212 +0x218
panic({0x67da00, 0xc000027968})
	/nix/store/njs424yyxxjhxibkpx5m36flr26nzzn6-go-1.17.8/share/go/src/runtime/panic.go:1038 +0x215
github.com/tektoncd/pipeline/pkg/git.checkLogMessage({0x696596, 0x12}, 0x6e4e00, 0x1, 0xc0001a7860)
	/home/vincent/src/tektoncd/pipeline/pkg/git/git_test.go:421 +0x1a5
github.com/tektoncd/pipeline/pkg/git.TestFetch.func1(0xc0001a7860)
	/home/vincent/src/tektoncd/pipeline/pkg/git/git_test.go:369 +0x4e5
testing.tRunner(0xc0001a7860, 0xc000079a10)
	/nix/store/njs424yyxxjhxibkpx5m36flr26nzzn6-go-1.17.8/share/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/nix/store/njs424yyxxjhxibkpx5m36flr26nzzn6-go-1.17.8/share/go/src/testing/testing.go:1306 +0x35a
FAIL	github.com/tektoncd/pipeline/pkg/git	0.212s

```

Ideally, we should set a temporary git config when running the test.

# Steps to Reproduce the Problem

1. Set `$HOME/.git/config` (or `$HOME/.config/git/config`) as a read-only file
2. Run `go test ./pkg/git`
3. See it fail

/assign

--- a/pkg/git/git_test.go
+++ b/pkg/git/git_test.go
@@ -29,7 +29,31 @@ import (
 
 const fileMode = 0755 // rwxr-xr-x
 
+func withTemporaryGitConfig(t *testing.T) func() {
+	gitConfigDir := t.TempDir()
+	key := "GIT_CONFIG_GLOBAL"
+	t.Helper()
+	oldValue, envVarExists := os.LookupEnv(key)
+	if err := os.Setenv(key, filepath.Join(gitConfigDir, "config")); err != nil {
+		t.Fatal(err)
+	}
+	clean := func() {
+		t.Helper()
+		if !envVarExists {
+			if err := os.Unsetenv(key); err != nil {
+				t.Fatal(err)
+			}
+			return
+		}
+		if err := os.Setenv(key, oldValue); err != nil {
+			t.Fatal(err)
+		}
+	}
+	return clean
+}
+
 func TestValidateGitSSHURLFormat(t *testing.T) {
+	withTemporaryGitConfig(t)
 	tests := []struct {
 		url  string
 		want bool
@@ -117,6 +141,7 @@ func TestValidateGitSSHURLFormat(t *testing.T) {
 }
 
 func TestValidateGitAuth(t *testing.T) {
+	withTemporaryGitConfig(t)
 	tests := []struct {
 		name       string
 		url        string
@@ -167,6 +192,7 @@ func TestValidateGitAuth(t *testing.T) {
 }
 
 func TestUserHasKnownHostsFile(t *testing.T) {
+	withTemporaryGitConfig(t)
 	tests := []struct {
 		name               string
 		want               bool
@@ -204,6 +230,7 @@ func TestUserHasKnownHostsFile(t *testing.T) {
 }
 
 func TestEnsureHomeEnv(t *testing.T) {
+	withTemporaryGitConfig(t)
 	tests := []struct {
 		name                 string
 		homeenvSet           bool
@@ -257,6 +284,7 @@ func TestEnsureHomeEnv(t *testing.T) {
 }
 
 func TestFetch(t *testing.T) {
+	withTemporaryGitConfig(t)
 	tests := []struct {
 		name       string
 		logMessage string

